### PR TITLE
add logs to execution

### PR DESCRIPTION
**Pull request by Liaojinghui**

### CODE REVIEW

**Comment by Liaojinghui** [comment]
will update utest later

**Review by shargon — COMMENTED** [review]
Could you do it without the neo changes? Attach an event before and remove later, like in neocUT

**Comment by Liaojinghui** [comment]
> Could you do it without the neo changes? Attach an event before and remove later, like in neo UT

@shargon working on that

File: src/RpcServer/RpcServer.SmartContract.cs [hunk 1]
```diff
@@ -83,6 +83,14 @@ private JObject GetInvokeResult(byte[] script, Signer[] signers = null, Witness[
                     obj["state"] = ToJson(n.State, session);
                     return obj;
                 }));
+                json["logs"] = new JArray(session.Engine.Notifications.Select(n =>
```
Comment (erikzhang) [inline src/RpcServer/RpcServer.SmartContract.cs:86]: This is incorrect.

File: src/ApplicationLogs/LogReader.cs [hunk 2]
```diff
@@ -54,6 +59,19 @@ protected override void OnSystemLoaded(NeoSystem system)
...
+            {
+                if (e.ScriptContainer is not Transaction tx) return;
+                if (!_logEvents.TryGetValue(tx.Hash, out var list))
+                {
+                    list = new();
+                    _logEvents.Add(tx.Hash, list);
+                }
+                list.Add(e);
+            }
+            ApplicationEngine.Log += Ev;
```
Comment (erikzhang) [inline src/ApplicationLogs/LogReader.cs:74]: Should we add a setting in config.json to control whether logs are saved or not? I think maybe not all nodes want to do this.

Comment (Liaojinghui) [reply src/ApplicationLogs/LogReader.cs:74]: The term `log` and `notify` is abusively miss used, I don't think the user could tell which is log and which is notify.

File: src/RpcServer/RpcServer.SmartContract.cs (hunk 1 above)
Comment (Liaojinghui) [reply src/RpcServer/RpcServer.SmartContract.cs:86]: This logic was when i try to add the log in the core, checking and removing useless code.

**Comment by Liaojinghui** [comment]
@nicolegys  May you please help me test this?

File: src/ApplicationLogs/LogReader.cs [hunk 3]
```diff
@@ -176,8 +176,13 @@ private void OnCommitting(NeoSystem system, Block block, DataCache snapshot, IRe
             foreach (var appExec in applicationExecutedList.Where(p => p.Transaction != null))
             {
                 var txJson = TxLogToJson(appExec);
+                if (RpcServerPlugin.LogEvents.TryGetValue(appExec.Transaction.Hash, out var list))
+                {
+                    txJson["logs"] = list.Select(q => q.Message) as JArray;
```
Comment (erikzhang) [inline src/ApplicationLogs/LogReader.cs:181]: I don't think `as` works.

File: src/RpcServer/RpcServerPlugin.cs [hunk 4]
```diff
@@ -54,6 +65,34 @@ protected override void OnSystemLoaded(NeoSystem system)
...
+                    _logs = new Queue<LogEventArgs>();
+                    LogEvents.Add(tx.Hash, _logs);
+                }
+                if (LogEvents[tx.Hash].Count >= MaxLogEvents)
+                {
+                    _logs.Dequeue();
+                }
+                _logs.Enqueue(e);
+            }
+            ApplicationEngine.Log += Ev;
```
Comment (erikzhang) [inline src/RpcServer/RpcServerPlugin.cs:86]: This method will be called multiple times.

**Comment by devhawk** [comment]
Why are we doing this?

File: src/ApplicationLogs/LogReader.cs [hunk 5]
```diff
@@ -25,7 +25,7 @@ namespace Neo.Plugins
 {
     public class LogReader : Plugin
     {
-        private DB db;
+        private DB _db;
```
Comment (devhawk) [inline src/ApplicationLogs/LogReader.cs:28]: Changes like this make it much harder to review PRs. Reviewers need to check each change to see if it is substantive or name change.

File: src/RpcServer/RpcServer.csproj [hunk 6]
```diff
@@ -4,5 +4,9 @@
     <PackageId>Neo.Plugins.RpcServer</PackageId>
     <RootNamespace>Neo.Plugins</RootNamespace>
   </PropertyGroup>
+
+  <ItemGroup>
+    <ProjectReference Include="..\RpcClient\RpcClient.csproj" />
```
Comment (devhawk) [inline src/RpcServer/RpcServer.csproj:9]: Why did we add this reference?

File: src/RpcServer/RpcServerPlugin.cs [hunk 7]
```diff
@@ -54,6 +65,36 @@ protected override void OnSystemLoaded(NeoSystem system)
...
             servers.TryAdd(s.Network, server);
+
+            ApplicationEngine.Log += Ev;
+            Blockchain.Committed += OnCommitted;
+        }
+
+        // It is potentially possible to have dos attack by sending a lot of transactions and logs.
+        // To prevent this, we limit the number of logs to be logged per contract.
+        // If the number of logs is greater than MAX_LOG_EVENTS, we remove the oldest log.
+        private static void Ev(object _, LogEventArgs e)
```
Comment (devhawk) [inline src/RpcServer/RpcServerPlugin.cs:76]: All of this log tracking logic should be in RpcServer not RpcServerPlugin.

The separation of RpcServerPlugin and RpcServer is to enable NeoExpress to host RpcServer directly. So any changes to behavior of this plugin must be in RpcServer class so that NeoExpress can use it

File: src/RpcServer/RpcServer.cs [hunk 8]
```diff
@@ -49,6 +48,7 @@ public RpcServer(NeoSystem system, RpcServerSettings settings)
             Initialize_SmartContract();
         }
 
+
```
Comment (devhawk) [inline src/RpcServer/RpcServer.cs:51]: Like the variable name change in LogReader.cs, these kind of changes make it harder to review PRs. We should do changes like this in a dedicated PR like https://github.com/neo-project/neo/pull/2785

**Review by devhawk — DISMISSED** [review]
Log tracking logic needs to be moved into RpcServer

File: src/RpcServer/RpcServer.csproj (hunk 6 above)
Comment (Liaojinghui) [reply src/RpcServer/RpcServer.csproj:9]: This is a  mistake, will remove it.

File: src/ApplicationLogs/LogReader.cs (hunk 5 above)
Comment (Liaojinghui) [reply src/ApplicationLogs/LogReader.cs:28]: This pr contains no substantive change, its just typos, comments, name styles. I open this pr simply because i found some  typos when i was going to work on  another feature.

**Comment by Liaojinghui** [comment]
@devhawk done as you suggested

File: src/RpcServer/RpcServerPlugin.cs [hunk 9]
```diff
@@ -56,6 +52,13 @@ protected override void OnSystemLoaded(NeoSystem system)
             servers.TryAdd(s.Network, server);
         }
 
+        public override void Dispose()
```
Comment (devhawk) [inline src/RpcServer/RpcServerPlugin.cs:55]: It looks like we could revert all the changes to this file in this PR

File: src/RpcServer/RpcServer.cs [hunk 10]
```diff
@@ -47,6 +58,14 @@ public RpcServer(NeoSystem system, RpcServerSettings settings)
...
             RegisterMethods(this);
             Initialize_SmartContract();
+
+            ApplicationEngine.Log += Ev;
+            Blockchain.Committed += OnCommitted;
+        }
+
+        private void OnCommitted(NeoSystem system, Block block)
+        {
+            LogEvents.Clear();
```
Comment (devhawk) [inline src/RpcServer/RpcServer.cs:68]: Should we be clearing all the events or just the ones for the transactions in `block`?

File: src/RpcServer/RpcServer.cs [hunk 11]
```diff
@@ -92,8 +111,28 @@ private static JObject CreateResponse(JObject id)
             return response;
         }
 
+        // It is potentially possible to have dos attack by sending a lot of transactions and logs.
+        // To prevent this, we limit the number of logs to be logged per contract.
+        // If the number of logs is greater than MAX_LOG_EVENTS, we remove the oldest log.
+        private static void Ev(object _, LogEventArgs e)
```
Comment (devhawk) [inline src/RpcServer/RpcServer.cs:117]: Can we change `Ev` to something more descriptive like `OnLog`

File: src/ApplicationLogs/LogReader.cs [hunk 12]
```diff
@@ -176,8 +176,13 @@ private void OnCommitting(NeoSystem system, Block block, DataCache snapshot, IRe
             foreach (var appExec in applicationExecutedList.Where(p => p.Transaction != null))
             {
                 var txJson = TxLogToJson(appExec);
+                if (RpcServer.LogEvents.TryGetValue(appExec.Transaction.Hash, out var list))
+                {
+                    txJson["logs"] = new JArray(list.Select(q => new JString(q.Message)));
+                }
                 Put(appExec.Transaction.Hash.ToArray(), Neo.Utility.StrictUTF8.GetBytes(txJson.ToString()));
             }
+            RpcServer.LogEvents.Clear();
```
Comment (erikzhang) [inline src/ApplicationLogs/LogReader.cs:185]: `Clear()` has been called in RpcServer.

Comment (Liaojinghui) [reply src/ApplicationLogs/LogReader.cs:185]: This is to make sure logs being removed from the memory as soon as possible.

File: src/RpcServer/RpcServer.cs (hunk 10 above)
Comment (Liaojinghui) [reply src/RpcServer/RpcServer.cs:68]: I think we should clear all log events, otherwise some logs will be there forever, which will be a memory leak.

Comment (erikzhang) [reply src/RpcServer/RpcServer.cs:68]: If you clear them in `OnCommitting`, other plugins that depend on this data may not work. I think it should be cleaned up in `OnCommitted` by a certain plugin.

File: src/RpcServer/RpcServer.cs [hunk 13]
```diff
@@ -39,6 +41,15 @@ public partial class RpcServer : IDisposable
...
         private readonly LocalNode localNode;
+        /// <summary>
+        /// Public interface of _logEvents.
+        /// </summary>
+        public static Dictionary<UInt256, Queue<LogEventArgs>> LogEvents { get; } = new();
+
+        /// <summary>
+        /// Maximum number of events to be logged per contract
+        /// </summary>
+        private const int MaxLogEvents = 50;
```
[the remark this answers is not in the document]
Comment (nicolegys) [reply src/RpcServer/RpcServer.cs:53]: +1 @Liaojinghui

**Comment by nicolegys** [comment]
Test pass

**Comment by Liaojinghui** [comment]
@shargon May you please review this pr?

**Comment by superboyiii** [comment]
@shargon @erikzhang Could you review it?

File: src/RpcServer/RpcServer.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (shargon) [reply src/RpcServer/RpcServer.cs:53]: Agree with @devhawk
